src/niftyone/figures/anat.py

```diff
@@ -0,0 +1,47 @@
+"""Generate figures for participant raw pipeline."""
+
+import logging
+from pathlib import Path
+
+import nibabel as nib
+import pandas as pd
+from bids2table.entities import BIDSEntities
+
+import niclips.image as noimg
+from niclips.figures import multi_view
+from niftyone import metrics
+
+
+def raw_t1w(
```

Comment:
Leaving a note from our offline conversation. Currently two figures plus qc metrics are being generated here. We discussed decoupling this a bit more. Specifically, we discussed:

1. Separating the concerns of figure generation and QC metric extraction.
2. Uncoupling the figures so that there is one generator per figure
3. Assigning the responsibility of filtering the records to the generator itself. (Currently that responsibility lies outside in the calling pipeline.)

One possible abstraction we discussed was:

```python
class Generator:
    def __init__(self, filters: dict[str, Any]):
        self.filters = filters

    def __call__(self, table: BIDSTable, out_dir: Path, overwrite: bool = False):
        """Filter table and generate figures for each matching record."""
        table = table.filter_multi(**self.filters)
        for _, record in table.iterrows():
            self.generate(record, out_dir, overwrite)

    def generate(self, record: pd.Series, out_dir: Path, overwrite: bool = False):
        """Generate figure for record."""
        raise NotImplementedError


class Pipeline:
    def __init__(self, generators: list[Generator]):
        self.generators = generators

    def run(self, table: BIDSTable, out_dir: Path, overwrite: bool = False):
        for gen in self.generators:
            gen(table, out_dir, overwrite)
```

For configuring pipelines, we could define the list of figure generators (and their kwargs) in a yaml config. Then at runtime, construct the list of generators and run the pipeline.

```yaml
figures:
  three_view:
    filters:
      suffix: T1w
  slice_video:
    filters:
      suffix: T1w
  three_view_video:
    filters:
      suffix: bold
  carpet_plot:
    filters:
      suffix: bold
  mean_std:
    filters:
      suffix: bold
```

The generator lookup could be handled with a global registry, with generators registered via a decorator when the class is defined.

I'm not fully convinced this is the best abstraction, so I'm open to alternatives. In particular, the config looks quite repetitive. This is not necessarily the worst thing, since decoupling is worth some repetition. But it might be a sign we are missing a better idea. 

Maybe something along the lines of letting groups of figures be associated with each set of BIDS `filters`, rather than a 1-1 mapping of filters to figures.

@nx10 thoughts?

Reply:
Hmm, another (not fully flushed out) thought if going the yaml route:

```yaml
key1:
  filters:
      desc: [True, brain, False]
      suffix: [T1w, T2w, T1map]
  views: [three_view, slice_video]

key2:
  filters:
      suffix: bold
  views: [three_view_video, carpet_plot, mean_std]
```

The entities provided in filters are zipped and passed along together. If an entity `True`, it treats that entity as a wildcard (e.g), if `False`, exclude rows that contain an entity for that filter. Otherwise, the filter has to match exactly. If an entity isn't provided, we can also treat it similar to `False`. So in the above case, it would generate the views for `sub-<subject>_desc-*_T1w`, `sub-<subject>_desc-brain_T2w`, and `sub-<subject>_T1map`.

The views will indicate the figure(s) that should be generated for those matching filters. Figure generation time aside, the end-user can always ignore those in NiftyOne itself. The primary keys (e.g. `key1`, `key2`) are kind of arbitrary and really are only there to group filters with figures to generate. 

As an aside, we're going to need some good documentation so that a user can find the different views that can be generated (irrespective of the implementation method) 😅 .

And also open to other ideas of course.

Reply:
Overall I would prefer a [Pandas query string](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html) or [Polars SQL query](https://docs.pola.rs/py-polars/html/reference/api/polars.SQLContext.execute.html) (for SQL I would just let the user set `SELECT * FROM df WHERE {userinput}`). 

It's always tempting to reimplement a small query DSL because one thinks they don't need all of that complexity, but then slowly realize that the complexity is needed.

Reply:
With @kaitj s matrix groups this would look like:

```yaml
my_t1w_images:
  query: "( desc in (None, 'brain') ) and (suffix in ('T1w', 'T2w', 'T1map') )"
  views: [three_view, slice_video]

my_bold_images:
  query: "suffix == 'bold'"
  views: [three_view_video, carpet_plot, mean_std]
```

Group name could also become a tag, then the same logic can be used to generate user defined tags (potentially without views):

```yaml
really_bad_images:
  query: "whatever_qc_metric < 0"
```

Reply:
Thanks @kaitj, @nx10 for the good suggestions. Let's go with your guys' preferred direction. I also think having a `query` option is a good idea. Having a simple way to do basic filters might also be a good idea, but it's less clear to me what the api for this should be. So I think let's just implement the `query` option for now and decide later on if there are common verbose query patterns we would like to streamline.

The suggestion for QC metric queries is also interesting. On the one hand, I can imagine a use case for generating figures conditional on some QC value. E.g. certain expensive to generate detailed figures might only be useful for bad or borderline data. On the other, it could be a bit of an anti-pattern since filtering by QC values can happen naturally within the fiftyone app. Also, it would require joining the QC metric table(s) with the b2t table prior to figure generation. (Not a big deal but something to note.) @kaitj what do you think about QC based queries?

Reply:
Yeah, I like the `query` option too. I think it'll be familiar too (hopefully) for what I imagine would be the majority of users.

Regarding the QC metric queries, I think we are on a similar page. I can see a use case for it too. In fact, it is likely what we would want to do for identifying problematic images as you noted, but I also think it is probably better to leave that kind of filteirng for within the app or potentially for power(?) users (e.g. [interactive plot](https://docs.voxel51.com/user_guide/plots.html) sort of thing).